File: Runtime/CppRefl.Tests/Source/Tests/MetadataTests.cpp
#include "gtest/gtest.h"

#include "MetadataCode.h"

#if TEST_METADATA_CODE()

#include "Reflection/ClassInfo.h"
#include "Reflection/FieldInfo.h"

TEST(MetadataTests, Tags)
{
	EXPECT_EQ(ClassWithNoMetadata::StaticReflectedClass().mTags.size(), 0);
	EXPECT_EQ(ClassWithNoMetadata::StaticReflectedClass().mTags.begin(), ClassWithNoMetadata::StaticReflectedClass().mTags.end());

	EXPECT_EQ(ClassWithMetadata::StaticReflectedClass().mTags.size(), 2);
	EXPECT_EQ(ClassWithMetadata::StaticReflectedClass().mTags[0], cpprefl::Name("TagNoSpaces"));
	EXPECT_EQ(ClassWithMetadata::StaticReflectedClass().mTags[1], cpprefl::Name("Tag Spaces"));

	EXPECT_EQ(ClassWithMetadata::StaticReflectedClass().mFields[0].mTags.size(), 1);
	EXPECT_EQ(ClassWithMetadata::StaticReflectedClass().mFields[0].mTags[0], cpprefl::Name("Field Tag"));
}

TEST(MetadataTests, Attributes)
{
	EXPECT_EQ(ClassWithNoMetadata::StaticReflectedClass().mAttributes.size(), 0);
	EXPECT_EQ(ClassWithNoMetadata::StaticReflectedClass().mAttributes.begin(), ClassWithNoMetadata::StaticReflectedClass().mAttributes.end());

	EXPECT_EQ(ClassWithMetadata::StaticReflectedClass().mAttributes.size(), 5);
	EXPECT_EQ(ClassWithMetadata::StaticReflectedClass().mAttributes[0].first, cpprefl::Name("Attr"));
	EXPECT_EQ(ClassWithMetadata::StaticReflectedClass().mAttributes[0].second, cpprefl::Name("Value"));
	EXPECT_EQ(ClassWithMetadata::StaticReflectedClass().mAttributes[1].first, cpprefl::Name("Attr 2"));
	EXPECT_EQ(ClassWithMetadata::StaticReflectedClass().mAttributes[1].second, cpprefl::Name("Value 2"));
	EXPECT_EQ(ClassWithMetadata::StaticReflectedClass().mAttributes[2].first, cpprefl::Name("Attr Number"));
	EXPECT_EQ(ClassWithMetadata::StaticReflectedClass().mAttributes[2].second.mInt, 12);
	EXPECT_EQ(ClassWithMetadata::StaticReflectedClass().mAttributes[3].first, cpprefl::Name("Attr Unquoted String"));
	EXPECT_EQ(ClassWithMetadata::StaticReflectedClass().mAttributes[3].second, cpprefl::Name("ClassWithNoMetadata"));
	EXPECT_EQ(ClassWithMetadata::StaticReflectedClass().mAttributes[4].first, cpprefl::Name("Attr Multiple"));
	EXPECT_EQ(ClassWithMetadata::StaticReflectedClass().mAttributes[4].second, cpprefl::Name("1, 2, 3"));

	EXPECT_EQ(ClassWithMetadata::StaticReflectedClass().mFields[0].mAttributes.size(), 1);
	EXPECT_EQ(ClassWithMetadata::StaticReflectedClass().mFields[0].mAttributes[0].first, cpprefl::Name("Field Attr"));
	EXPECT_EQ(ClassWithMetadata::StaticReflectedClass().mFields[0].mAttributes[0].second, cpprefl::Name("Value"));
}

#endif

File: Compiler/CppRefl.Compiler.Tests/Reflection/MetadataTests.cs
using CppRefl.Compiler.Reflection;

namespace CppRefl.Compiler.Tests.Reflection
{
    internal class MetadataTests : CompilerTestBase
    {
        protected override string ReflectionFile => Path.Combine(Const.Runtime.Tests.ReflectedHeadersDirectory.FullName, "MetadataCode.h");

        private void ValidateTag(MetadataValue tag, string expectedTag, MetadataValueLifetime expectedLifetime)
        {
            Assert.That(tag.Value, Is.EqualTo(expectedTag));
            Assert.That(tag.Lifetime, Is.EqualTo(expectedLifetime));
        }

        private void ValidateAttribute(MetadataValue attribute, string expectedValue, MetadataValueLifetime expectedLifetime)
        {
            Assert.That(attribute.Value, Is.EqualTo(expectedValue));
            Assert.That(attribute.Lifetime, Is.EqualTo(expectedLifetime));
        }

        [Test]
        public void ClassTags()
        {
            var classInfo = Registry.GetClass("ClassWithMetadata")!;

            Assert.That(classInfo.Metadata.Tags.Count, Is.EqualTo(3));
            ValidateTag(classInfo.Metadata.Tags[0], "TagNoSpaces", MetadataValueLifetime.Runtime);
            ValidateTag(classInfo.Metadata.Tags[1], "Tag Spaces", MetadataValueLifetime.Runtime);
            ValidateTag(classInfo.Metadata.Tags[2], "Compile Only Tag", MetadataValueLifetime.Compile);
        }

        [Test]
        public void ClassAttributes()
        {
            var classInfo = Registry.GetClass("ClassWithMetadata")!;

            Assert.That(classInfo.Metadata.Attributes.Count, Is.EqualTo(6));
            ValidateAttribute(classInfo.Metadata.Attributes["Attr"], "\"Value\"", MetadataValueLifetime.Runtime);
            ValidateAttribute(classInfo.Metadata.Attributes["Attr 2"], "\"Value 2\"", MetadataValueLifetime.Runtime);
            ValidateAttribute(classInfo.Metadata.Attributes["Attr Number"], "12", MetadataValueLifetime.Runtime);
            ValidateAttribute(classInfo.Metadata.Attributes["Attr Unquoted String"], "ClassWithNoMetadata", MetadataValueLifetime.Runtime);
            ValidateAttribute(classInfo.Metadata.Attributes["Attr Multiple"], "1, 2, 3", MetadataValueLifetime.Runtime);
            ValidateAttribute(classInfo.Metadata.Attributes["Compile Only Attr"], "\"Compile Only Value\"", MetadataValueLifetime.Compile);
        }

        [Test]
        public void FieldMetadata()
        {
            var classInfo = Registry.GetClass("ClassWithMetadata")!;
            var fieldInfo = classInfo.Fields[0];

            Assert.That(fieldInfo.Metadata.Tags.Count, Is.EqualTo(2));
            ValidateTag(fieldInfo.Metadata.Tags[0], "Field Tag", MetadataValueLifetime.Runtime);
            ValidateTag(fieldInfo.Metadata.Tags[1], "Compile Tag", MetadataValueLifetime.Compile);

            Assert.That(fieldInfo.Metadata.Attributes.Count, Is.EqualTo(2));
            ValidateAttribute(fieldInfo.Metadata.Attributes["Field Attr"], "\"Value\"", MetadataValueLifetime.Runtime);
            ValidateAttribute(fieldInfo.Metadata.Attributes["Compile Attr"], "\"Compile Value\"", MetadataValueLifetime.Compile);
        }

        [Test]
        public void MethodMetadata()
        {
            var classInfo = Registry.GetClass("ClassWithMetadata")!;
            var methodInfo = classInfo.Methods[0];

            Assert.That(methodInfo.Metadata.Tags.Count, Is.EqualTo(2));
            ValidateTag(methodInfo.Metadata.Tags[0], "Method Tag", MetadataValueLifetime.Runtime);
            ValidateTag(methodInfo.Metadata.Tags[1], "Compile Tag", MetadataValueLifetime.Compile);

            Assert.That(methodInfo.Metadata.Attributes.Count, Is.EqualTo(2));
            ValidateAttribute(methodInfo.Metadata.Attributes["Method Attr"], "\"Value\"", MetadataValueLifetime.Runtime);
            ValidateAttribute(methodInfo.Metadata.Attributes["Compile Attr"], "\"Compile Value\"", MetadataValueLifetime.Compile);
        }

        [Test]
        public void SourceLocation()
        {
            var classInfo = Registry.GetClass("ClassWithMetadata")!;
            Assert.That(classInfo.Metadata.SourceLocation.FileInfo.FullName, Is.EqualTo(ReflectionFile));
            Assert.That(classInfo.Metadata.SourceLocation.Line, Is.EqualTo(28));
            Assert.That(classInfo.GeneratedBodyLine, Is.EqualTo(30));

            var enumInfo = Registry.GetEnum("EnumWithMetadata")!;
            Assert.That(enumInfo.Metadata.SourceLocation.FileInfo.FullName, Is.EqualTo(ReflectionFile));
            Assert.That(enumInfo.Metadata.SourceLocation.Line, Is.EqualTo(47));
            Assert.That(enumInfo.GeneratedBodyLine, Is.EqualTo(49));
        }

        [Test]
        public void CommentTests()
        {
            // No comments
            {
                var classInfo = Registry.GetClass("ClassWithNoMetadata")!;
                Assert.IsNull(classInfo.Metadata.Comment);
            }

            // With comments
            {
                var classInfo = Registry.GetClass("ClassWithMetadata")!;
                Assert.That(classInfo.Metadata.Comment, Is.EqualTo("Class comment."));
                Assert.That(classInfo.Fields[0].Metadata.Comment, Is.EqualTo("Field comment."));
                Assert.That(classInfo.Methods[0].Metadata.Comment, Is.EqualTo("Multline function comment."));
            }

        }
    }
}